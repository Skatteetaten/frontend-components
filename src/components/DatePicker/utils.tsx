export const DEFAULTFORMATDATE = (date: Date | null | undefined): string => {
  return date
    ? Intl.DateTimeFormat('no', {
        year: 'numeric',
        month: '2-digit',
        day: 'numeric',
      }).format(date)
    : '';
};

export const DEFAULTPARSEDATEFROMSTRING = (date: string): Date | null => {
  try {
    if (date && date.length === 10) {
      const day: number = parseInt(date.substring(0, 2));
      const month: number = parseInt(date.substring(3, 5));
      const year: number = parseInt(date.substring(6, 10));
      const formattedDate: Date = new Date(Date.UTC(year, month - 1, day));
      return formattedDate;
    }
  } catch (error) {
    console.log(error);
  }
  return null;
};

export const monthsForLocale = (
  localeName = 'no',
  monthFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { month: monthFormat })
    .format;
  const months: string[] = [];
  for (let month = 0; month < 12; month++) {
    months.push(format(new Date(Date.UTC(2020, month))));
  }
  return months;
};

export const weekdaysForLocale = (
  localeName = 'no',
  dayFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { weekday: dayFormat })
    .format;
  const weekdays: string[] = [];
  for (let day = 0; day < 7; day++) {
    weekdays.push(format(new Date(Date.UTC(2020, 7, day + 2)))); // add 2 days to let the week start on Sunday
  }
  return weekdays;
};
