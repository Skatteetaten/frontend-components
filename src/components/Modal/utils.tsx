import { useEffect } from 'react';

export const createModalDomPlacement = (
  setFlag: (a: boolean) => void,
  shadowRootNode: Document | ShadowRoot | undefined
) => {
  if (!shadowRootNode) {
    setFlag(true);
    return;
  }

  const styleInjectorWrapperId = 'style-injector';
  const modalWrapperId = 'modal-wrapper';

  if (shadowRootNode.getElementById(modalWrapperId)) {
    setFlag(true);
  } else {
    const modalWrapper = document.createElement('div');
    modalWrapper.id = modalWrapperId;
    const styleInjectorWrapper = shadowRootNode.getElementById(
      styleInjectorWrapperId
    );
    const parent = styleInjectorWrapper ? styleInjectorWrapper : shadowRootNode;
    parent.appendChild(modalWrapper);
    setFlag(true);
  }
};

export const useEscOnPress = (
  ref: HTMLDivElement | null,
  onClose?: (ref: HTMLDivElement) => void
): void => {
  useEffect(() => {
    const onKeyDown = (event: KeyboardEvent): void => {
      if (ref == null) {
        return;
      }
      if (event.code === 'Escape' && onClose) {
        onClose(ref);
      }
    };

    if (ref) {
      window.addEventListener('keydown', onKeyDown as EventListener, false);
    }
    return function cleanup(): void {
      if (window) {
        window.removeEventListener('keydown', onKeyDown as EventListener);
      }
    };
  }, [ref, onClose]);
};
