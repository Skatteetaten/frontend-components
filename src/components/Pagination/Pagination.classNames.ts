import { getTheme } from '@uifabric/styling';
import { PaletteProps } from '../../../lib';
import { mergeStyleSets } from '@uifabric/merge-styles';

export const getClassNames = () => {
  const palette = getTheme().palette as PaletteProps;

  return mergeStyleSets({
    activePage: {
      color: `${palette.skeColor.darkGrey} !important`,
      borderBottom: 'none !important',
      paddingTop: '3px !important'
    },
    linkIcons: {
      fontSize: '22px',
      verticalAlign: 'middle'
    },
    pageNumber: {
      borderBottom: `2px solid ${palette.skeColor.lightBlue} !important`,
      paddingTop: '3px !important'
    },
    paginationContainer: {
      fontSize: '16px',
      selectors: {
        button: {
          border: 'none',
          background: 'transparent',
          fontSize: '16px'
        },
        nav: {
          float: 'right',
          selectors: {
            '@media only screen and (max-width: 479px)': {
              margin: '0 auto',
              float: 'none'
            }
          }
        },
        p: {
          marginTop: '7px',
          float: 'left',
          selectors: {
            '@media only screen and (max-width: 479px)': {
              margin: '0 auto',
              float: 'none'
            }
          }
        },
        ul: {
          listStyle: 'none',
          display: 'inline-flex',
          paddingLeft: '0',
          margin: 0
        },
        li: {
          display: 'inline',
          selectors: {
            button: {
              color: palette.skeColor.blue,
              fontWeight: 'bold',
              padding: '0',
              margin: '4px',
              cursor: 'pointer'
            }
          }
        },
        '@media only screen and (max-width: 479px)': {
          display: 'grid'
        }
      }
    }
  });
};
