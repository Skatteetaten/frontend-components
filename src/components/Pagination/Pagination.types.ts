export interface PaginationProps {
  /** Aria-label */
  ariaLabel?: string;
  /** Aria-label for sidelinker */
  ariaLabelNavigationLink?: string;
  /** Aria-label for aktiv sidelink */
  ariaLabelNavigationLinkActive?: string;
  /** CSS class */
  className?: string;
  /** Nåværende sidenummer */
  currentPage: number;
  /** Tekst neste-knapp */
  nextLabel?: string;
  /** Funksjon som trigges på sideendring */
  onPageChange: (page: number) => void;
  /** Antall sidetall som vises samtidig. Default er 3. */
  pagesDisplayed?: number;
  /** Total antall objekter i liste */
  total: number;
  /** Ant objekter som vises per side */
  pageSize: number;
  /** Tekst forrige-knapp */
  previousLabel?: string;
}
