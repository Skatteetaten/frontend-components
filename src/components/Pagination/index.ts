export { Pagination as default } from './Pagination';
export * from './Pagination';
