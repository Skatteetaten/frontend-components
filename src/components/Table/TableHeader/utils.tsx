import { TFunction } from 'react-i18next';

export const getIconNameForTheadCell = (
  isSorted: boolean,
  isSortedAscending: boolean
) => {
  if (isSorted) {
    return isSortedAscending ? 'ArrowDown' : 'ArrowUp';
  }

  return 'ArrowUpDown';
};

export const getAriaLabelForTheadCell = (
  name: string | JSX.Element,
  fieldName: string,
  isSorted: boolean,
  isSortedAscending: boolean,
  t: TFunction<Array<string>>
) => {
  let ariaLabel = fieldName;
  if (typeof name === 'string') {
    ariaLabel = isSorted
      ? isSortedAscending
        ? name.concat(' ', t('table.sorted_ascending'))
        : name.concat(' ', t('table.sorted_descending'))
      : name.concat(' ', t('table.sortable'));
  }

  return ariaLabel;
};
