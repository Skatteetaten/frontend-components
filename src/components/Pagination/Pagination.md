** Pagination **

```js
import Table from '@skatteetaten/frontend-components/Table';
import Pagination from '@skatteetaten/frontend-components/Pagination';
const columns = [
  {
    name: 'Navn',
    fieldName: 'navn'
  },
  {
    name: 'Tilgang gitt',
    fieldName: 'dato',
    alignment: 'right'
  },
  {
    name: 'Type',
    fieldName: 'type',
    alignment: 'right'
  }
];
const data = [
  {
    navn: 'Kaisa Lundquist',
    dato: '23.05.16',
    type: 'Bruker'
  },
  {
    navn: 'Reidar Olsen',
    dato: '15.11.16',
    type: 'Bruker'
  },
  {
    navn: 'Bob Egil Hansen',
    dato: '15.11.16',
    type: 'Bruker'
  },
  {
    navn: 'Leif KÃ¥re Lund',
    dato: '23.10.17',
    type: 'Admin'
  },
  {
    navn: 'Kai Mossige',
    dato: '25.11.19',
    type: 'Bruker'
  },
  {
    navn: 'Bob Testman',
    dato: '01.01.20',
    type: 'Bruker'
  },
  {
    navn: 'Alice Middleman',
    dato: '25.03.20',
    type: 'Admin'
  },
  {
    navn: 'Kari Saksbehandler',
    dato: '25.03.20',
    type: 'Admin'
  },
  {
    navn: 'Frank Johansen',
    dato: '26.03.20',
    type: 'Bruker'
  },
  {
    navn: 'Tallulah Willis',
    dato: '01.05.20',
    type: 'Admin'
  }
];
const [displayedData, setDisplayedData] = React.useState(
  [...data].splice(0, 3)
);
const [currentPage, setCurrentPage] = React.useState(1);
const pageSize = 3;

<div>
  <Table data={displayedData} columns={columns} />
  <Pagination
    currentPage={currentPage}
    onPageChange={page => {
      const index = (page - 1) * pageSize;
      setDisplayedData([...data].splice(index, 3));
      setCurrentPage(page);
    }}
    total={data.length}
    pageSize={pageSize}
  />
</div>;
```
