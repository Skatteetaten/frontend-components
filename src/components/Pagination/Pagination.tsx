import * as React from 'react';
import { getClassNames } from './Pagination.classNames';
import classnames from 'classnames';
import Icon from '../Icon';

interface PaginationProps {
  /** Aria-label */
  ariaLabel?: string;
  /** Aria-label for sidelinker */
  ariaLabelNavigationLink?: string;
  /** Aria-label for aktiv sidelink */
  ariaLabelNavigationLinkActive?: string;
  /** CSS class */
  className: string;
  /** Nåværende sidenummer */
  currentPage: number;
  /** Tekst neste-knapp */
  nextLabel?: string;
  /** Funksjon som trigges på sideendring */
  onPageChange: (page: number) => void;
  /** Antall sidetall som vises samtidig. Default er 3. */
  pagesDisplayed?: number;
  /** Total antall objekter i liste */
  total: number;
  /** Ant objekter som vises per side */
  pageSize: number;
  /** Tekst forrige-knapp */
  previousLabel?: string;
}
const getNumberOfPages = (total: number, pageSize: number) => {
  return Math.ceil(total / pageSize);
};

const range = (start: number, end: number, pagesDisplayed: number) =>
  Array.from(
    { length: pagesDisplayed < end ? pagesDisplayed : end },
    (v, k) => k + start
  );

export const getSlidingWindowEdges = (
  currentPage: number,
  total: number,
  pageSize: number,
  pagesDisplayed: number
) => {
  const numberOfPages = getNumberOfPages(total, pageSize);
  if (numberOfPages <= pagesDisplayed) {
    return {
      startPage: 1,
      endPage: numberOfPages
    };
  }
  let startPage = currentPage - (pagesDisplayed - 1);
  if (currentPage <= pagesDisplayed) {
    startPage = 1;
  }
  let endPage = startPage + (pagesDisplayed - 1);

  if (endPage > numberOfPages) {
    endPage = numberOfPages;
  }
  return {
    startPage: startPage,
    endPage: endPage
  };
};

const NextPage: React.FC<{
  currentPage: number;
  total: number;
  pageSize: number;
  onClick: (newPage: number) => void;
  label: string;
}> = props => {
  const styles = getClassNames();

  return (
    <li>
      <button
        onClick={evt => {
          evt.preventDefault();
          props.onClick(props.currentPage + 1);
        }}
        role="link"
      >
        <span>{props.label}</span>
        <Icon iconName="ChevronRight" className={styles.linkIcons} />
      </button>
    </li>
  );
};

const PreviousLink: React.FC<{
  currentPage: number;
  onClick: (newPage: number) => void;
  label: string;
}> = props => {
  const styles = getClassNames();
  return (
    <li>
      <button
        onClick={evt => {
          evt.preventDefault();
          props.onClick(props.currentPage - 1);
        }}
        role="link"
      >
        <Icon iconName="ChevronLeft" className={styles.linkIcons} />
        {props.label}
      </button>
    </li>
  );
};

const Page: React.FC<{
  page: number;
  onClick: (page: number) => void;
  isCurrent: boolean;
  ariaLabelNavigationLink: string | undefined;
  ariaLabelNavigationLinkActive: string | undefined;
}> = props => {
  const {
    page,
    onClick,
    isCurrent,
    ariaLabelNavigationLink,
    ariaLabelNavigationLinkActive
  } = props;
  const styles = getClassNames();

  const ariaLabel = () => {
    if (isCurrent) {
      return (
        (ariaLabelNavigationLinkActive
          ? ariaLabelNavigationLinkActive
          : 'Side ') + page
      );
    }
    return (ariaLabelNavigationLink ? ariaLabelNavigationLink : 'Side ') + page;
  };

  return (
    <li>
      <button
        onClick={evt => {
          evt.preventDefault();
          onClick(page);
        }}
        className={isCurrent ? styles.activePage : styles.pageNumber}
        aria-label={ariaLabel()}
        aria-current={isCurrent}
        aria-disabled={isCurrent}
        role="link"
      >
        {page}
      </button>
    </li>
  );
};

const Pages = (props: {
  currentPage: number;
  pagesDisplayed: number | undefined;
  pageSize: number;
  onPageChange: (newPage: number) => void;
  total: number;
  ariaLabelNavigationLink: string | undefined;
  ariaLabelNavigationLinkActive: string | undefined;
}) => {
  const pagesDisplayed = props.pagesDisplayed || 3;
  const windowEdges = getSlidingWindowEdges(
    props.currentPage,
    props.total,
    props.pageSize,
    pagesDisplayed
  );

  return (
    <div>
      {range(windowEdges.startPage, windowEdges.endPage, pagesDisplayed).map(
        i => {
          return (
            <Page
              key={i}
              onClick={props.onPageChange}
              page={i}
              isCurrent={props.currentPage === i}
              ariaLabelNavigationLink={props.ariaLabelNavigationLink}
              ariaLabelNavigationLinkActive={
                props.ariaLabelNavigationLinkActive
              }
            />
          );
        }
      )}
    </div>
  );
};

/**
 * @visibleName Pagination
 */
const Pagination: React.FC<PaginationProps> = props => {
  const {
    ariaLabel,
    ariaLabelNavigationLink,
    ariaLabelNavigationLinkActive,
    className,
    currentPage,
    nextLabel,
    onPageChange,
    pagesDisplayed,
    pageSize,
    previousLabel,
    total
  } = props;
  const styles = getClassNames();
  const firstListObject = (currentPage - 1) * pageSize;
  const lastListObject = firstListObject + pageSize;
  const view =
    lastListObject > total ? total : firstListObject + 1 + '-' + lastListObject;
  return (
    <div className={classnames(styles.paginationContainer, className)}>
      <p>
        Viser {view} av {total}
      </p>
      <nav aria-label={ariaLabel ? ariaLabel : 'Sidenavigering'}>
        <ul>
          {currentPage > 1 && (
            <PreviousLink
              currentPage={currentPage}
              onClick={(page: number) => onPageChange(page)}
              label={previousLabel ? previousLabel : 'Forrige'}
            />
          )}
          <li>
            <ul>
              <Pages
                onPageChange={onPageChange}
                currentPage={currentPage}
                total={total}
                pagesDisplayed={pagesDisplayed}
                pageSize={pageSize}
                ariaLabelNavigationLink={ariaLabelNavigationLink}
                ariaLabelNavigationLinkActive={ariaLabelNavigationLinkActive}
              />
            </ul>
          </li>
          {currentPage < getNumberOfPages(total, pageSize) && (
            <NextPage
              currentPage={currentPage}
              total={total}
              pageSize={pageSize}
              onClick={onPageChange}
              label={nextLabel ? nextLabel : 'Neste'}
            />
          )}
        </ul>
      </nav>
    </div>
  );
};

export default Pagination;
