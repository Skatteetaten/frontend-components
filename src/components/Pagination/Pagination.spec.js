import React from 'react';
import Pagination, { getSlidingWindowEdges } from './Pagination';
import { matches } from '../utils/test-utils';
import { mount } from 'enzyme';

function oppsettFullDOM(props) {
  const Wrapper = wrapperProps => {
    const [currentPage, setCurrentPage] = React.useState(1);
    return (
      <Pagination
        {...wrapperProps}
        currentPage={currentPage}
        onPageChange={page => setCurrentPage(page)}
      />
    );
  };

  return mount(<Wrapper {...props} />);
}

describe('Pagination komponent', () => {
  it('matcher snapshot', () => {
    matches(
      <Pagination
        currentPage={1}
        onPageChange={() => jest.fn()}
        total={15}
        pageSize={5}
      />
    );
  });
  it('skal rendre i henhold til props', () => {
    const wrapper = oppsettFullDOM({
      ariaLabelNavigationLink: 'Go to page ',
      ariaLabelNavigationLinkActive: 'Current page, Page ',
      total: 15,
      pageSize: 5
    });
    const firstPageLink = wrapper
      .find('ul')
      .last()
      .find('li')
      .first()
      .find('button');
    const lastPageLink = wrapper
      .find('ul')
      .last()
      .find('li')
      .last()
      .find('button');
    expect(firstPageLink.prop('aria-label')).toEqual('Current page, Page 1');
    expect(firstPageLink.prop('aria-current')).toEqual(true);
    expect(lastPageLink.prop('aria-label')).toEqual('Go to page 3');
    expect(lastPageLink.prop('aria-current')).toEqual(false);
  });
  it('skal endre aktiv side på klikk', () => {
    const wrapper = oppsettFullDOM({
      total: 15,
      pageSize: 5
    });
    const firstPageLink = wrapper
      .find('ul')
      .last()
      .find('li')
      .first()
      .find('button');
    const lastPageLink = wrapper
      .find('ul')
      .last()
      .find('li')
      .last()
      .find('button');
    expect(firstPageLink.prop('aria-current')).toEqual(true);
    expect(lastPageLink.prop('aria-current')).toEqual(false);

    wrapper
      .find('ul')
      .last()
      .find('li')
      .last()
      .find('button')
      .simulate('click');
    expect(
      wrapper
        .find('ul')
        .last()
        .find('li')
        .first()
        .find('button')
        .prop('aria-current')
    ).toEqual(false);
    expect(
      wrapper
        .find('ul')
        .last()
        .find('li')
        .last()
        .find('button')
        .prop('aria-current')
    ).toEqual(true);
  });
  it('getSlidingWindowEdges() skal resturne riktig start og slutt basert på input', () => {
    let range = getSlidingWindowEdges(1, 12, 4, 2);
    expect(range.startPage).toEqual(1);
    expect(range.endPage).toEqual(2);
    range = getSlidingWindowEdges(16, 100, 5, 5);
    expect(range.startPage).toEqual(12);
    expect(range.endPage).toEqual(16);
  });
  it('skal genere en side dersom total er mindre enn pageSize', () => {
    let wrapper = oppsettFullDOM({
      total: 17,
      pageSize: 50
    });
    expect(
      wrapper
        .find('ul')
        .last()
        .find('li').length
    ).toEqual(1);
    wrapper = oppsettFullDOM({
      total: 32,
      pageSize: 20
    });
    expect(
      wrapper
        .find('ul')
        .last()
        .find('li').length
    ).toEqual(2);
    expect(
      wrapper
        .find('ul')
        .last()
        .find('li')
        .first()
        .find('button')
        .prop('aria-current')
    ).toEqual(true);
  });
});
