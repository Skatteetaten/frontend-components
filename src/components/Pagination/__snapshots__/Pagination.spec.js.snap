// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination komponent matcher snapshot 1`] = `
<div
  className=

      {
        font-size: 1rem;
        padding-top: 24px;
      }
      & button {
        background: transparent;
        border: none;
        font-size: 1rem;
      }
      & nav {
        float: right;
      }
      @media only screen and (max-width: 479px){& nav {
        float: none;
        margin-bottom: 0;
        margin-left: auto;
        margin-right: auto;
        margin-top: 0;
      }
      & p {
        float: left;
        margin-top: 7px;
      }
      @media only screen and (max-width: 479px){& p {
        float: none;
        margin-bottom: 0;
        margin-left: auto;
        margin-right: auto;
        margin-top: 0;
      }
      & ul {
        display: inline-flex;
        list-style: none;
        margin-bottom: 0px;
        margin-left: 0px;
        margin-right: 0px;
        margin-top: 0px;
        padding-left: 0;
      }
      & li {
        display: inline;
      }
      & li button {
        color: #1362ae;
        cursor: pointer;
        font-weight: bold;
        margin-bottom: 4px;
        margin-left: 6px;
        margin-right: 6px;
        margin-top: 4px;
        padding-bottom: 3px;
        padding-left: 1px;
        padding-right: 1px;
        padding-top: 3px;
        vertical-align: middle;
      }
      & li button:focus {
        background-color: #cde1f9;
        color: #093e61;
        outline: none;
      }
      & li button:hover {
        background-color: #cde1f9;
        color: #093e61;
      }
      @media only screen and (max-width: 479px){& {
        display: grid;
      }
>
  <p
    data-testid="pagination-oppsummering"
  >
    Viser 1-5 av 15
  </p>
  <nav
    aria-label="Sidenavigering"
  >
    <ul>
      <li>
        <ul>
          <div>
            <li>
              <button
                aria-current={true}
                aria-disabled={true}
                aria-label="Side 1"
                className=

                    {
                      border-bottom: none !important;
                      color: #4d4d4d !important;
                      padding-bottom: 6px !important;
                      padding-top: 3px !important;
                    }
                onClick={[Function]}
                role="link"
              >
                1
              </button>
            </li>
            <li>
              <button
                aria-current={false}
                aria-disabled={false}
                aria-label="Side 2"
                className=

                    {
                      border-bottom: 2px solid #cde1f9 !important;
                      padding-top: 3px !important;
                    }
                onClick={[Function]}
                role="link"
              >
                2
              </button>
            </li>
            <li>
              <button
                aria-current={false}
                aria-disabled={false}
                aria-label="Side 3"
                className=

                    {
                      border-bottom: 2px solid #cde1f9 !important;
                      padding-top: 3px !important;
                    }
                onClick={[Function]}
                role="link"
              >
                3
              </button>
            </li>
          </div>
        </ul>
      </li>
      <li>
        <button
          onClick={[Function]}
          role="link"
        >
          <span>
            Neste
          </span>
          <i
            aria-hidden={true}
            className=

                {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  display: inline-block;
                  font-family: "SkeIconfont";
                  font-size: 1.375rem;
                  font-style: normal;
                  font-weight: normal;
                  speak: none;
                  text-rendering: auto;
                  transform: translate(0, 0);
                  vertical-align: middle;
                }
            data-icon-name="ChevronRight"
          >
            î¨ž
          </i>
        </button>
      </li>
    </ul>
  </nav>
</div>
`;
